File: benchmarks/binarytrees/ref.rb
# binarytrees reference, Ruby. Two-element arrays; nil is a leaf.
def build(d) = d.zero? ? nil : [build(d - 1), build(d - 1)]
def check(t) = t.nil? ? 1 : 1 + check(t[0]) + check(t[1])

maxdepth = (ARGV[0] || "14").to_i
longlived = build(maxdepth)
d = 4
while d <= maxdepth
  iters = 1 << (maxdepth - d + 4)
  acc = 0
  iters.times { acc += check(build(d)) }
  puts "#{iters} trees of depth #{d} check #{acc}"
  d += 2
end
puts "long-lived tree of depth #{maxdepth} check #{check(longlived)}"

File: benchmarks/binarytrees/ref.js
// binarytrees reference, Node. Two-element arrays; null is a leaf.
function build(d) { return d === 0 ? null : [build(d - 1), build(d - 1)]; }
function check(t) { return t === null ? 1 : 1 + check(t[0]) + check(t[1]); }

const maxdepth = parseInt(process.argv[2] || "14", 10);
const longlived = build(maxdepth);
for (let d = 4; d <= maxdepth; d += 2) {
  const iters = 2 ** (maxdepth - d + 4);
  let acc = 0;
  for (let i = 0; i < iters; i++) acc += check(build(d));
  console.log(iters + " trees of depth " + d + " check " + acc);
}
console.log("long-lived tree of depth " + maxdepth + " check " + check(longlived));
